File: libs/moqtail-rs/src/model/parameter/switch_transition.rs
// Copyright 2025 The MOQtail Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//! SWITCH_TRANSITION (moq-transport PR #1378, "SWITCH for Client-side ABR").
//!
//! Carried on the target Track's PUBLISH a relay opens in answer to a SWITCH so
//! the subscriber learns where the seam is:
//!
//! ```text
//! SWITCH_TRANSITION {
//!   Switching Group ID (i),   // G_switch: first group on the new track
//!   Live Edge Group ID (i)    // target's live edge when the PUBLISH was opened
//! }
//! ```
//!
//! Wire-wise it is the odd-typed (0x73) bytes-valued message parameter
//! [`MessageParameter::SwitchTransition`]; this module holds the typed value and
//! its two-varint payload codec.

use bytes::{Buf, Bytes, BytesMut};

use crate::model::common::varint::{BufMutVarIntExt, BufVarIntExt};
use crate::model::error::ParseError;
use crate::model::parameter::message_parameter::MessageParameter;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct SwitchTransition {
  /// G_switch: the first group_id delivered on the target Track.
  pub switching_group_id: u64,
  /// The target Track's live edge group_id at the moment the PUBLISH opened.
  pub live_edge_group_id: u64,
}

impl SwitchTransition {
  pub fn new(switching_group_id: u64, live_edge_group_id: u64) -> Self {
    Self {
      switching_group_id,
      live_edge_group_id,
    }
  }

  /// The typed message parameter ready to drop into a PUBLISH's parameter list.
  pub fn to_message_parameter(&self) -> MessageParameter {
    MessageParameter::SwitchTransition {
      switching_group_id: self.switching_group_id,
      live_edge_group_id: self.live_edge_group_id,
    }
  }

  /// Encode the two-varint payload of a SWITCH_TRANSITION value.
  pub fn to_bytes(&self) -> Result<Bytes, ParseError> {
    let mut payload = BytesMut::new();
    payload.put_vi(self.switching_group_id)?;
    payload.put_vi(self.live_edge_group_id)?;
    Ok(payload.freeze())
  }

  /// Decode the two-varint payload of a SWITCH_TRANSITION value.
  pub fn from_bytes(mut value: Bytes) -> Result<Self, ParseError> {
    let switching_group_id = value.get_vi()?;
    let live_edge_group_id = value.get_vi()?;
    if value.has_remaining() {
      return Err(ParseError::KeyValueFormattingError {
        context: "SwitchTransition::from_bytes",
      });
    }
    Ok(Self {
      switching_group_id,
      live_edge_group_id,
    })
  }

  /// Find the first SWITCH_TRANSITION parameter in a list, if any.
  pub fn from_parameters(params: &[MessageParameter]) -> Option<Self> {
    params.iter().find_map(|p| match p {
      MessageParameter::SwitchTransition {
        switching_group_id,
        live_edge_group_id,
      } => Some(Self::new(*switching_group_id, *live_edge_group_id)),
      _ => None,
    })
  }
}

#[cfg(test)]
mod tests {
  use super::*;
  use crate::model::common::pair::KeyValuePair;
  use crate::model::parameter::constant::MessageParameterType;

  #[test]
  fn payload_roundtrip() {
    let st = SwitchTransition::new(42, 100);
    let bytes = st.to_bytes().unwrap();
    assert_eq!(SwitchTransition::from_bytes(bytes).unwrap(), st);
  }

  #[test]
  fn roundtrip_via_message_parameter_wire_form() {
    // Full KeyValuePair serialize/deserialize cycle, as it travels on a PUBLISH
    // parameter list: an odd type, so a length-prefixed bytes value.
    let st = SwitchTransition::new(7, 9);
    let kvp: KeyValuePair = st.to_message_parameter().try_into().unwrap();
    assert_eq!(
      kvp.get_type(),
      MessageParameterType::SwitchTransition as u64
    );
    let mut buf = kvp.serialize().unwrap();
    let parsed = KeyValuePair::deserialize(&mut buf).unwrap();
    let param = MessageParameter::deserialize(&parsed).unwrap();
    assert_eq!(SwitchTransition::from_parameters(&[param]).unwrap(), st);
  }

  #[test]
  fn from_parameters_ignores_other_params() {
    let other = MessageParameter::new_delay_groups(5);
    assert!(SwitchTransition::from_parameters(&[other]).is_none());
    assert!(SwitchTransition::from_parameters(&[]).is_none());
  }

  #[test]
  fn trailing_bytes_are_rejected() {
    let mut payload = BytesMut::new();
    payload.put_vi(1).unwrap();
    payload.put_vi(2).unwrap();
    payload.put_vi(3).unwrap();
    assert!(SwitchTransition::from_bytes(payload.freeze()).is_err());
  }

  #[test]
  fn large_group_ids_roundtrip() {
    let st = SwitchTransition::new(1_000_000, 2u64.pow(30));
    assert_eq!(
      SwitchTransition::from_bytes(st.to_bytes().unwrap()).unwrap(),
      st
    );
  }
}

File: libs/moqtail-ts/src/model/parameter/switch_transition.ts
/**
 * Copyright 2025 The MOQtail Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { ByteBuffer, FrozenByteBuffer } from '../common/byte_buffer'
import { KeyValuePair } from '../common/pair'
import { VersionSpecificParameterType } from './constant'

/**
 * SWITCH_TRANSITION version-specific parameter (draft-ietf-moq-transport
 * PR #1378, "SWITCH for Client-side ABR"). TS mirror of the Rust
 * `parameter::switch_transition::SwitchTransition`.
 *
 * The relay attaches this parameter to the target Track's PUBLISH so the
 * subscriber knows exactly where the seam is after a SWITCH:
 *
 * ```text
 * SWITCH_TRANSITION {
 *   Switching Group ID (i),   // G_switch — first group on the new track
 *   Live Edge Group ID (i)    // target's live edge when PUBLISH was opened
 * }
 * ```
 *
 * Everything in `[switchingGroupId, liveEdgeGroupId)` arrives on the catch-up
 * FETCH_HEADER stream; everything from the live edge onward arrives on the
 * normal SUBGROUP_HEADER streams. A subscriber can also use `switchingGroupId`
 * to discard buffered old-track content at or above the seam (the draft's
 * buffer-replacement use case).
 *
 * Wire-wise this is an *odd*-typed (`0x73`) bytes-valued {@link KeyValuePair}
 * whose value is the two varints back to back. On a *failure* PUBLISH
 * (content_exists = 0, immediately followed by PUBLISH_DONE) the relay still
 * carries this parameter to mark the PUBLISH as switch-related, with `{0, 0}`
 * as placeholder values — key off the PUBLISH_DONE status code, not these
 * values, in that case.
 */
export class SwitchTransition {
  constructor(
    /** G_switch: the first group_id delivered on the target Track. */
    public readonly switchingGroupId: bigint,
    /** The target Track's live edge group_id at the moment the PUBLISH opened. */
    public readonly liveEdgeGroupId: bigint,
  ) {}

  /** Encode as the `0x73` bytes-valued {@link KeyValuePair} ready to drop into a PUBLISH's parameter list. */
  toKeyValuePair(): KeyValuePair {
    const payload = new ByteBuffer()
    payload.putVI(this.switchingGroupId)
    payload.putVI(this.liveEdgeGroupId)
    return KeyValuePair.tryNewBytes(VersionSpecificParameterType.SwitchTransition, payload.toUint8Array())
  }

  /** Decode the two-varint payload of a SWITCH_TRANSITION value. Throws on malformed input. */
  static fromBytes(value: Uint8Array): SwitchTransition {
    const buf = new FrozenByteBuffer(value)
    const switchingGroupId = buf.getVI()
    const liveEdgeGroupId = buf.getVI()
    return new SwitchTransition(switchingGroupId, liveEdgeGroupId)
  }

  /**
   * Find and decode the first SWITCH_TRANSITION parameter in a list, if any.
   * Returns `null` when the parameter is absent or carries the wrong shape
   * (varint-valued, or a malformed byte payload).
   */
  static fromParameters(params: KeyValuePair[]): SwitchTransition | null {
    for (const p of params) {
      if (p.typeValue === BigInt(VersionSpecificParameterType.SwitchTransition) && p.value instanceof Uint8Array) {
        try {
          return SwitchTransition.fromBytes(p.value)
        } catch {
          return null
        }
      }
    }
    return null
  }
}